azurelinuxagent/common/rdma.py

```diff
@@ -265,11 +265,144 @@ def provision_network_direct_rdma(self):
+            return (None, None)
+
+        key_0 = None
+        value_0 = None
+        f = open(driver_info_source, "rb")  # pylint: disable=C0103
+        while True:
+            key = f.read(kvp_key_size)
+            value = f.read(kvp_value_size)
+            if key and value:
+                key_0 = key.partition(b"\x00")[0]
+                if key_0 :
+                    key_0 = key_0.decode()
+                if key_0 == "IPoIB_Data":
+                    value_0 = value.partition(b"\x00")[0]
+                    if value_0 :
+                        value_0 = value_0.decode()
+                    break
+            else:
+                break
+        f.close()
+
+        if key_0 == "IPoIB_Data":
+            return (key_0, value_0)
+
+        return (None, None)
+
+    def provision_sriov_multiple_ib(self, value) :
+
+        mac_ip_array = []
+
+        values = value.split("|")
+        num_ips = len(values) - 1
+        # values[0] tells how many IPs. Format - NUMPAIRS:<number>
+        match = re.match(r"NUMPAIRS:(\d+)", values[0])
+        if match:
+            num = int(match.groups(0)[0])
+            if num != num_ips:
+                logger.error("RDMA: multiple IPs reported num={0} actual number of IPs={1}".format(num, num_ips))
+                return
+        else:
+            logger.error("RDMA: failed to find number of IP addresses in {0}".format(values[0]))
+            return
+
+        for i in range(1, num_ips+1):
+            # each MAC/IP entry is of format <MAC>:<IP>
+            match = re.match(r"([^:]+):(\d+\.\d+\.\d+\.\d+)", values[i])
+            if match:
+                mac_addr = match.groups(0)[0]
+                ipv4_addr = match.groups(0)[1]
+                mac_ip_array.append((mac_addr, ipv4_addr))
+            else:
+                logger.error("RDMA: failed to find MAC/IP address in {0}".format(values[i]))
+                return
+
+        # try to assign all MAC/IP addresses to IB interfaces
+        # retry for up to 60 times, with 1 seconds delay between each
+        retry = 60
+        while retry > 0:
+            count = self.update_iboip_interfaces(mac_ip_array)
+            if count == len(mac_ip_array):
+                return
+
+            time.sleep(1)
+            retry -= 1
+
+        logger.error("RDMA: failed to set all IP over IB addresses")
+
+    # Assign addresses to all IP over IB interfaces specified in mac_ip_array
+    # Return the number of IP addresses successfully assigned
+
+    def update_iboip_interfaces(self, mac_ip_array):
+
+        net_dir = "/sys/class/net"
+        nics = os.listdir(net_dir)
+        count = 0
+
+        for nic in nics:
+            # look for IBoIP interface of format ibXXX
+            if not re.match(r"ib\d+", nic):
+                continue
+            address_file = open(net_dir + "/" + nic + "/address")
+            mac_addr = address_file.read()
+            address_file.close()
```

Comment:
Wrap this in a `finally` clause to ensure the file is always closed 
(Consider using `with open(filename)` instead as that ensures that the file is always closed even on error - Eg: https://cmdlinetips.com/2016/01/opening-a-file-in-python-using-with-statement/)